Rewrite `emoji_mart_data_light` as TS

nxxkap opened the pull request:
This PR rewrites `emoji_mart_data_light` as TS.

The imported module `emoji_compressed.js`, it was difficult to rewrite to TS due to the convenience of using `babel-plugin-preval`, so I took the form of creating a result type.

Please feel free to share your thoughts, feedback, or suggestions. Any advice for further improvement is highly appreciated.

Commit:
Format and rename to ts

Commit:
Add type

Commit:
Resolve ESLint `prefer-const` rule

ClearlyClaire commented:
Tagging @renchap, as iirc he wants to remove usage of `babel-plugin-preval` as part of his work moving away from webpack to ViteJS.

renchap commented:
I will have a look at this.

The move to ViteJS will still use `preval`, we need to rework the whole emoji loading to get rid of it.

As a quick glance, a few remarks:
- could you create a `emoji_compressed.d.ts` file and define its exported types here, rather than doing it in `emoji_mart_data_light.ts` ? This would be cleaner
- The file header says: 
  > The output of this module is designed to mimic emoji-mart's "data" object

  Could it be possible to use `emoji's mart` types here to define the export from their `data` property, so we know it is consistent?

Gargron commented:
Whatever preval is doing, couldn't it theoretically be moved to an npm action to generate the right file?

renchap commented:
> Whatever preval is doing, couldn't it theoretically be moved to an npm action to generate the right file?

Yes, it could, as it is generating the same files for each build. We could generate them and commit them to the repo for once.

I also think that this might not be useful anymore with modern bundlers / treeshaking / newer Emoji Picker, but this is a big refactor as this data is use at other places, so this is low on my backlog at the moment.

taichi221228 commented:
> could you create a `emoji_compressed.d.ts` file and define its exported types here, rather than doing it in `emoji_mart_data_light.ts` ? This would be cleaner

Thank you for your review. I will begin with separating the types.
In terms of separating the types, I'm unsure of the best way to proceed. When using `*.d.ts`, all defined types fall into the global scope. This could lead to potentially inappropriate naming, such as `FilenameData` and `Native`.

1. Should I define types in `*.d.ts` and change to more explicit type names like `EmojiFilenameData`?
2. Should I define types in `*.d.ts` and retain the current type names like `FilenameData`?
3. Should I define types in `.ts`, using names like `FilenameData` internally only, while creating and exporting more comprehensive types like `EmojiCompressed` and `Emojis`?

I would appreciate your advice on how best to proceed.

renchap commented:
You should be able to write a `emoji_compressed.d.ts` in the same folder as the `.js`, and use it to define the types for your `.js` file. It will not pollute the global scope:

```ts
// emoji_compressed.d.ts

export default const emojiCompressed: [
   ShortCodesToEmojiData,
   Skins,
   Category[],
   ShortName[]
 ];
```

Typescript doc on how to achieve this: https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html

Commit:
Add separate file for type definitions

Commit:
Add type annotations to emoji_compressed.js export

taichi221228 commented:
@renchap Thanks, I fixed the type definition.

Commit:
Fix comment

Commit:
Fix `FilenameData` type

Commit:
Fix omission of `Alias[]` type in `EmojiCompressed`

Commit:
Refactored to use types defined in `emoji-mart`

Commit:
Add clarifying comments for `Search` and `Skins` types in type definitions

This commit adds detailed comments for the `Search` and `Skins` types in the type definition file. The comments explain the current state of these types and highlight potential areas for future updates or refactoring.

Commit:
Document absence of `skins` property in `emoji_compressed.js`

This commit adds a comment to clarify the non-existence of the `skins` property in the `emoji_compressed.js` file. The comment includes links to related areas in the codebase where further information can be found.

Commit:
Fix comment

Commit:
Fix `FilenameData` type

Commit:
Fix array destructure order in `emojiCompressed` to pass tests

renchap commented on app/javascript/mastodon/features/emoji/emoji_mart_data_light.ts:
```diff
@@ -0,0 +1,55 @@
+  skins,
+  categories,
+  short_names,
+  emojisWithoutShortCodes, // eslint-disable-line @typescript-eslint/no-unused-vars
```
If you prefix the name with `_`, like `_emojisWithoutShortCodes`, then you dont need to ignore the line

taichi221228 replied:
Thanks, fixed it.

renchap commented on app/javascript/mastodon/features/emoji/emoji_mart_data_light.ts:
```diff
@@ -0,0 +1,55 @@
+// decompress
+Object.keys(shortCodesToEmojiData).forEach((shortCode) => {
+  const [
+    filenameData, // eslint-disable-line @typescript-eslint/no-unused-vars
```
Same as above

taichi221228 replied:
Thanks, fixed it.

renchap reviewed (changes requested):
2 small changes, but this looks good.

I assume you tested the various emoji-related feature in the UI to ensure everything still works?

Commit:
Remove unnecessary `@typescript-eslint/no-unused-vars`.

taichi221228 commented:
I've checked, albeit briefly, that the emoji-related features work as expected.

- I've tested whether emojis can be entered using the `:${emoji.id | short_names[0]}:` format, and verified that it doesn't distort the suggestion display.
- I've checked the functionality of the emoji picker, ensuring that emojis can be entered, searched for, and properly categorized without any display distortion or other operational issues.
- I've conducted these checks within the post functionality.
　
Please let me know if there are other areas you would like me to test or any issues you have noticed with the current implementation.

renchap reviewed (approved):
Thanks for working on this!
